hmrc-mongo-play-27/src/main/scala/uk/gov/hmrc/mongo/play/json/formats/MongoUuidFormats.scala

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2021 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.mongo.play.json.formats
+
+import play.api.libs.json._
+
+import java.nio.ByteBuffer
+import java.nio.ByteOrder
+import java.util.Base64
+import java.util.UUID
+
+trait MongoUuidFormats {
+  outer =>
+
+  private val encoder = Base64.getEncoder
+  private val decoder = Base64.getDecoder
+
+  private val UuidOldSubType = "03"
+  private val UuidNewSubtype = "04"
+
+  private def readBase64(str: String, byteOrder: ByteOrder): UUID = {
+    val bytes     = decoder.decode(str)
+    val buffer    = ByteBuffer.wrap(bytes).order(byteOrder)
+    val highBytes = buffer.getLong()
+    val lowBytes  = buffer.getLong()
+    new UUID(highBytes, lowBytes)
+  }
+
+  private def writeBase64(uuid: UUID, byteOrder: ByteOrder): String = {
+    val empty = new Array[Byte](16)
+    val bytes = ByteBuffer.wrap(empty).order(byteOrder)
+    bytes.putLong(uuid.getMostSignificantBits)
+    bytes.putLong(uuid.getLeastSignificantBits)
+    encoder.encodeToString(bytes.array())
+  }
+
+  final val uuidReads: Reads[UUID] =
+    Reads.at[String](__ \ "$binary" \ "subType").flatMap {
+      case `UuidOldSubType` =>
```

Comment:
Do we need to support the old sub type here? An alternative `MongoUuidFormat` could be provided for the old type, but I expect no-one has used the legacy format since it wasn't supported by hmrc-mongo before now.
Note, the registered `UuidCodec` is set to `Standard` above so they would be out of sync.

Reply:
Yeah, that was mostly out of an excess of caution - this currently supports reading both old and new format but will only write the new format.
The motivation is just that there's a possibility that teams will already have used the binary UUID via ReactiveMongo and might have been using it when the old subtype was the default - I don't know the history of ReactiveMongo well but it looks like it currently uses the STANDARD byte order as ByteBuffers default to BIG_ENDIAN:
https://github.com/ReactiveMongo/ReactiveMongo-BSON/blob/5ba467cd047403d586cfe7e5b15196739871192d/api/src/main/scala/DefaultBSONHandlers.scala#L239

Reply:
The problem would be that the registered `UuidCodec` only supports the `Standard` so depending on whether you are using the registered codec (filters/queries uuids) or the json format (reads/writes of top level entity) you would get different behaviour (codec would fail, json format wouldn't).
I think we should just support `Standard` or make it possible for the client to elect `Legacy` (and migrate themselves).

Reply:
I have split this into two traits now, so that the user can choose which subtype to write rather than only writing `STANDARD`